### PR TITLE
8337654: Relocate uncommon trap stub from SharedRuntime to OptoRuntime

### PR DESCRIPTION
**Pull request by adinn**
Reorganization of generation and management code for C2-specific blob so that it is, as far as possible, under the scope of class OptoRuntime with an implementation located in C2-specific source files.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337654](https://bugs.openjdk.org/browse/JDK-8337654): Relocate uncommon trap stub from SharedRuntime to OptoRuntime (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) 🔄 Re-review required (review applies to [7d7b3288](https://git.openjdk.org/jdk/pull/20417/files/7d7b3288edead11c6bcdf371fc6b9d3d084b07b9))
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) 🔄 Re-review required (review applies to [15fb5fa0](https://git.openjdk.org/jdk/pull/20417/files/15fb5fa0e2a865606e7613e94f3f07be2b5dd6be))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20417/head:pull/20417` \
`$ git checkout pull/20417`

Update a local copy of the PR: \
`$ git checkout pull/20417` \
`$ git pull https://git.openjdk.org/jdk.git pull/20417/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20417`

View PR using the GUI difftool: \
`$ git pr show -t 20417`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20417.diff">https://git.openjdk.org/jdk/pull/20417.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20417#issuecomment-2262832257)

### CODE REVIEW

**Comment by adinn** [comment]
> Did you forget to add new file runtime_risc.cpp to changeset?

Yes, apologies for that. Should be there now.

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/cpu/aarch64/runtime_aarch64.cpp [hunk 1]
```diff
@@ -36,6 +36,350 @@
...
+  // Use rfp because the frames look interpreted now
+  // Save "the_pc" since it cannot easily be retrieved using the last_java_SP after we aligned SP.
+  // Don't need the precise return PC here, just precise enough to point into this code blob.
+  address the_pc = __ pc();
+  __ set_last_Java_frame(sp, rfp, the_pc, rscratch1);
+
+  // Call C code.  Need thread but NOT official VM entry
+  // crud.  We cannot block on this call, no GC can happen.  Call should
+  // restore return values to their stack-slots with the new SP.
+  // Thread is in rdi already.
```
Comment (iwanowww) [inline src/hotspot/cpu/aarch64/runtime_aarch64.cpp:219]: It's weird to see `rdi` being mentioned in AArch64-specific code :-)

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by RealFYang** [comment]
Hi Andrew, I think we are lacking following addon change for riscv. Built fastdebug and manually tested on linux-riscv64.

```
diff --git a/src/hotspot/cpu/riscv/runtime_riscv.cpp b/src/hotspot/cpu/riscv/runtime_riscv.cpp
index 46fdcc65a2c..c030441857c 100644
--- a/src/hotspot/cpu/riscv/runtime_riscv.cpp
+++ b/src/hotspot/cpu/riscv/runtime_riscv.cpp
@@ -58,7 +58,7 @@ class SimpleRuntimeFrame {
 #define __ masm->

 //------------------------------generate_uncommon_trap_blob--------------------
-void SharedRuntime::generate_uncommon_trap_blob() {
+void OptoRuntime::generate_uncommon_trap_blob() {
   // Allocate space for the code
   ResourceMark rm;
   // Setup code generation tools
@@ -122,7 +122,7 @@ void SharedRuntime::generate_uncommon_trap_blob() {
     __ lwu(t0, Address(x14, Deoptimization::UnrollBlock::unpack_kind_offset()));
     __ mv(t1, Deoptimization::Unpack_uncommon_trap);
     __ beq(t0, t1, L);
-    __ stop("SharedRuntime::generate_uncommon_trap_blob: expected Unpack_uncommon_trap");
+    __ stop("OptoRuntime::generate_uncommon_trap_blob: expected Unpack_uncommon_trap");
     __ bind(L);
   }
 #endif
@@ -377,6 +377,5 @@ void OptoRuntime::generate_exception_blob() {
   // Set exception blob
   _exception_blob =  ExceptionBlob::create(&buffer, oop_maps, SimpleRuntimeFrame::framesize >> 1);
 }
-#endif // COMPILER2
-

+#endif // COMPILER2
```

**Comment by adinn** [comment]
Hi Fei (@RealFYang). Thanks for checking. I pushed a corrected version.

**Review by RealFYang — APPROVED** [review]
LGTM. Thanks.

File: src/hotspot/cpu/aarch64/runtime_aarch64.cpp (hunk 1 above)
Comment (adinn) [reply src/hotspot/cpu/aarch64/runtime_aarch64.cpp:219]: A hangover from the existing (relocated) code going back to when it was cloned from runtime_x86_64.cpp (where the comment still exists even though it appears also to be incorrect!). I'll delete it.

**Review by vnkozlov — APPROVED** [review]
Update is good.

**Pull request closed without merging**